src/cryptography/hazmat/primitives/ciphers/modes.py

```diff
@@ -104,90 +109,78 @@ def __init__(self, tweak):
 
-    def validate_for_algorithm(self, algorithm):
+    def validate_for_algorithm(self, algorithm: CipherAlgorithm):
         if algorithm.key_size not in (256, 512):
             raise ValueError(
                 "The XTS specification requires a 256-bit key for AES-128-XTS"
                 " and 512-bit key for AES-256-XTS"
             )
 
 
-@utils.register_interface(Mode)
-class ECB(object):
+class ECB(Mode):
     name = "ECB"
 
     validate_for_algorithm = _check_aes_key_length
 
 
-@utils.register_interface(Mode)
-@utils.register_interface(ModeWithInitializationVector)
-class OFB(object):
+class OFB(Mode, ModeWithInitializationVector):
     name = "OFB"
 
-    def __init__(self, initialization_vector):
+    def __init__(self, initialization_vector: bytes):
         utils._check_byteslike("initialization_vector", initialization_vector)
         self._initialization_vector = initialization_vector
 
     initialization_vector = utils.read_only_property("_initialization_vector")
     validate_for_algorithm = _check_iv_and_key_length
 
 
-@utils.register_interface(Mode)
-@utils.register_interface(ModeWithInitializationVector)
-class CFB(object):
+class CFB(Mode, ModeWithInitializationVector):
     name = "CFB"
 
-    def __init__(self, initialization_vector):
+    def __init__(self, initialization_vector: bytes):
         utils._check_byteslike("initialization_vector", initialization_vector)
         self._initialization_vector = initialization_vector
 
     initialization_vector = utils.read_only_property("_initialization_vector")
     validate_for_algorithm = _check_iv_and_key_length
 
 
-@utils.register_interface(Mode)
-@utils.register_interface(ModeWithInitializationVector)
-class CFB8(object):
+class CFB8(Mode, ModeWithInitializationVector):
     name = "CFB8"
 
-    def __init__(self, initialization_vector):
+    def __init__(self, initialization_vector: bytes):
         utils._check_byteslike("initialization_vector", initialization_vector)
         self._initialization_vector = initialization_vector
 
     initialization_vector = utils.read_only_property("_initialization_vector")
     validate_for_algorithm = _check_iv_and_key_length
 
 
-@utils.register_interface(Mode)
-@utils.register_interface(ModeWithNonce)
-class CTR(object):
+class CTR(Mode, ModeWithNonce):
     name = "CTR"
 
-    def __init__(self, nonce):
+    def __init__(self, nonce: bytes):
         utils._check_byteslike("nonce", nonce)
         self._nonce = nonce
 
     nonce = utils.read_only_property("_nonce")
 
-    def validate_for_algorithm(self, algorithm):
+    def validate_for_algorithm(self, algorithm: CipherAlgorithm):
         _check_aes_key_length(self, algorithm)
-        if len(self.nonce) * 8 != algorithm.block_size:
-            raise ValueError(
-                "Invalid nonce size ({}) for {}.".format(
-                    len(self.nonce), self.name
-                )
-            )
+        _check_nonce_length(self.nonce, self.name, algorithm)
```

Comment:
This is a workaround. mypy doesn't support intersection types and in this case we want to have an algorithm that is both a `CipherAlgorithm` and a `BlockCipherAlgorithm`. Since we can't have that we hide the call that requires `BlockCipherAlgorithm` behind an untyped call.

There may be a better way to do this.